plugins/inputs/http_listener/http_listener.go

```diff
@@ -158,6 +151,126 @@ func (t *HttpListener) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if req.ContentLength > t.MaxBodySize {
+		toolarge(res)
+		return
+	}
+
+	// Handle gzip request bodies
+	var body io.ReadCloser
+	if req.Header.Get("Content-Encoding") == "gzip" {
+		r, err := gzip.NewReader(req.Body)
+		defer r.Close()
+		if err != nil {
+			log.Println("E! " + err.Error())
+			badrequest(res)
+			return
+		}
+		body = http.MaxBytesReader(res, r, t.MaxBodySize)
+	} else {
+		body = http.MaxBytesReader(res, req.Body, t.MaxBodySize)
+	}
+
+	var return400 bool
+	var buf []byte
+	bufstart := 0
+	for {
+		if bufstart == 0 {
+			buf = t.pool.get()
+		}
+		n, err := io.ReadFull(body, buf[bufstart:])
+
+		if err != nil && err != io.ErrUnexpectedEOF && err != io.EOF {
+			log.Println("E! " + err.Error())
+			// problem reading the request body
+			badrequest(res)
+			t.pool.put(buf)
+			return
+		}
+
+		if err == io.ErrUnexpectedEOF || err == io.EOF || n == 0 {
+			// finished reading the request body
+			if err := t.parse(buf[:n+bufstart]); err != nil {
+				log.Println("E! " + err.Error())
+				return400 = true
+			}
+			t.pool.put(buf)
+			if return400 {
+				badrequest(res)
+			} else {
+				res.WriteHeader(http.StatusNoContent)
+			}
+			return
+		}
+
+		// if we got down here it means that we filled our buffer, and there
+		// are still bytes remaining to be read. So we will parse up until the
+		// final newline, then push the rest of the bytes into the next buffer.
+		i := bytes.LastIndexByte(buf, '\n')
+		if i == -1 {
+			// drop any line longer than the max buffer size
+			t.pool.put(buf)
+			newlinei := findnewline(body)
+			log.Printf("E! http_listener received a single line of %d bytes, maximum is %d bytes",
+				MAX_LINE_SIZE+newlinei, MAX_LINE_SIZE)
+			return400 = true
+			continue
+		}
+		if err := t.parse(buf[:i]); err != nil {
+			log.Println("E! " + err.Error())
+			return400 = true
+		}
+		// rotate the bit remaining after the last newline to the front of the buffer
+		bufstart = len(buf) - i
+		copy(buf[:bufstart], buf[i:])
+	}
+}
+
+func (t *HttpListener) parse(b []byte) error {
+	metrics, err := t.parser.Parse(b)
+
+	for _, m := range metrics {
+		t.acc.AddFields(m.Name(), m.Fields(), m.Tags(), m.Time())
+	}
+
+	return err
+}
+
+func toolarge(res http.ResponseWriter) {
+	res.Header().Set("Content-Type", "application/json")
+	res.Header().Set("X-Influxdb-Version", "1.0")
+	res.WriteHeader(http.StatusRequestEntityTooLarge)
+	res.Write([]byte(`{"error":"http: request body too large"}`))
+}
+
+func badrequest(res http.ResponseWriter) {
+	res.Header().Set("Content-Type", "application/json")
+	res.Header().Set("X-Influxdb-Version", "1.0")
+	res.WriteHeader(http.StatusBadRequest)
+	res.Write([]byte(`{"error":"http: bad request"}`))
+}
+
+// findnewline finds the next newline in the given reader. It returns the number
+// of bytes it had to read to get there.
+func findnewline(r io.Reader) int {
+	counter := 0
+	// read until the next newline:
+	var tmp [1]byte
```

Comment:
Reading one byte at a time (especially if it results in a syscall) is very inefficient. This might be better captured with some state in the main parsing loop. `var badLineBytes int`, and if that's greater than zero, scan for a newline in the next chunk, log the error, and drop those bytes.